src/prof_recent.c

```diff
@@ -0,0 +1,407 @@
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), tctx->tdata->lock);
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	/*
+	 * Increment recent_count to hold the tctx so that it won't be gone
+	 * even after tctx->tdata->lock is released.  This acts as a
+	 * "placeholder"; the real recording of the allocation requires a lock
+	 * on prof_recent_alloc_mtx and is done in prof_recent_alloc_slow (when
+	 * tctx->tdata->lock has been released).
+	 */
+	++tctx->recent_count;
+}
+
+static void
+decrement_recent_count(tsd_t *tsd, prof_tctx_t *tctx) {
+	malloc_mutex_assert_not_owner(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	if (tctx == NULL) {
+		return;
+	}
+	malloc_mutex_lock(tsd_tsdn(tsd), tctx->tdata->lock);
+	assert(tctx->recent_count > 0);
+	--tctx->recent_count;
+	prof_tctx_try_destroy(tsd, tctx);
+}
+
+#ifndef JEMALLOC_JET
+static
+#endif
+prof_recent_t *
+extent_prof_recent_alloc_get(const extent_t *extent) {
+	malloc_mutex_assert_owner(TSDN_NULL, &prof_recent_alloc_mtx);
+	prof_recent_t *recent_alloc = extent->e_prof_info.e_prof_recent_alloc;
+	assert(recent_alloc == NULL || recent_alloc->alloc_extent == extent);
+	return recent_alloc;
+}
+
+static void
+extent_prof_recent_alloc_set_unchecked(extent_t *extent,
+    prof_recent_t *recent_alloc) {
+	malloc_mutex_assert_owner(TSDN_NULL, &prof_recent_alloc_mtx);
+	assert(recent_alloc != NULL);
+	extent->e_prof_info.e_prof_recent_alloc = recent_alloc;
+	recent_alloc->alloc_extent = extent;
+}
+
+static void
+extent_prof_recent_alloc_reset(extent_t *extent, prof_recent_t *recent_alloc) {
+	malloc_mutex_assert_owner(TSDN_NULL, &prof_recent_alloc_mtx);
+	assert(recent_alloc != NULL);
+	assert(recent_alloc == extent_prof_recent_alloc_get(extent));
+	extent->e_prof_info.e_prof_recent_alloc = NULL;
+	recent_alloc->alloc_extent = NULL;
+}
+
+static void
+extent_prof_recent_alloc_reset_unchecked(extent_t *extent) {
+	malloc_mutex_assert_owner(TSDN_NULL, &prof_recent_alloc_mtx);
+	extent->e_prof_info.e_prof_recent_alloc = NULL;
+}
+
+/*
+ * This function should be called right before an allocation is released, so
+ * that the associated recent allocation record can contain the following
+ * information:
+ * (1) The allocation is released;
+ * (2) The time of the deallocation; and
+ * (3) The prof_tctx associated with the deallocation.
+ */
+void
+prof_recent_alloc_reset(tsd_t *tsd, extent_t *extent) {
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	prof_recent_t *recent = extent_prof_recent_alloc_get(extent);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	if (recent == NULL) {
+		return;
+	}
+
+	prof_tctx_t *dalloc_tctx = prof_tctx_create(tsd);
+	if (dalloc_tctx != NULL) {
+		malloc_mutex_lock(tsd_tsdn(tsd), dalloc_tctx->tdata->lock);
+		++dalloc_tctx->recent_count;
+		dalloc_tctx->prepared = false;
+		malloc_mutex_unlock(tsd_tsdn(tsd), dalloc_tctx->tdata->lock);
+	}
+
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
```

Comment:
Can we try making this function not taking the same lock twice? It should help simplify the logic.

Reply:
It looks the recent_alloc lock is lower ranked than tdata lock? We should lock from the beginning then.